feat: Advanced Authorization (#39)
eclipse-che/blog: _posts/2024-01-29-advanced-authorization.adoc

---
title: Advanced Authorization
layout: post
author: Ilya Buziuk
description: >-
  The article describes how to configure advanced authorization in Eclipse Che
categories: []

> keywords: ['postgresql']

Reviewer: Copy/paste?

slug: /@ilya.buziuk/advanced-authorization
---

> Even though the current trend is instead of having a gigantic monolith {kubernetes} cluster split the infrastructure into a number of 'Fit For Purpose' clusters, administrators still want to provide granular access and restrict certain functionality for particular users.

Reviewer 2: ```suggestion
Even though the current trend is to split an infrastructure into a number of 'Fit For Purpose' clusters instead of having a gigantic monolith {kubernetes} cluster, administrators still want to provide granular access and restrict certain functionality for particular users.
```

> NOTE: 'Fit For Purpose' {kubernetes} cluster refers to a cluster specifically designed and configured to meet the requirements and goals of a particular use case or workload, tailored to optimize performance, resource utilization, and other factors based on the characteristics of the workloads it will be managing. For {prod} it is recommended to have this type of cluster provisioned.

Reviewer 2: ```suggestion
NOTE: A 'Fit For Purpose' {kubernetes} cluster refers to a cluster specifically designed and configured to meet the requirements and goals of a particular use case or workload, tailored to optimize performance, resource utilization, and other factors based on the characteristics of the workloads it will be managing. For {prod} it is recommended to have this type of cluster provisioned.
```

> Previously, all users provisioned on a {kubernetes} cluster had access to {prod}, and it was problematic to restrict the usage for certain accounts. One of the options for administrators was to remove permissions for namespace / project creation on the {kubernetes} level, and disable automatic namespace provisioning for {prod}:

Reviewer 2: ```suggestion
Previously, all users provisioned on a {kubernetes} cluster had access to {prod}, and it was difficult to restrict the usage for certain accounts. One of the options for administrators was to remove permissions for namespace / project creation on the {kubernetes} level, and disable automatic namespace provisioning for {prod}:
```

[source,yaml]
----
devEnvironments:
  defaultNamespace:
    autoProvision: false
----

Considering this, new optional properties that allow setting up granular access for different groups and users have been added to the {prod} Custom Resource: `allowUsers`, `allowGroups`, `denyUsers`, and `denyGroups`. Below you can find an illustration exemplifying the configuration:

[source,yaml]
----
 networking:
    auth:
      advancedAuthorization:
        allowUsers:
          - user-a
          - user-b
        denyUsers:
          - user-c
        allowGroups:
          - team-a
          - team-b
        denyGroups:
          - team-c
----

IMPORTANT: `AllowGroups` and `DenyGroups` properties are only supported on OpenShift.

Users who are not allowed will not be able to use {prod} and will see the following warning when trying to access the User Dashboard:

image::/assets/img/advanced-authorization/warning.png[User might not be allowed warning]
{nbsp} +

== Dogfooding In Action

> For the development and dogfooding purposes internally we use link:https://www.redhat.com/en/technologies/cloud-computing/openshift/dedicated[OpenShift Dedicated] cluster with GitHub identity provider setup. Under the `Users` tab, all the accounts that have access to the cluster are listed e.g.

Reviewer 2: ```suggestion
For internal development and dogfooding purposes we use an link:https://www.redhat.com/en/technologies/cloud-computing/openshift/dedicated[OpenShift Dedicated] cluster with GitHub identity provider setup. Under the `Users` tab, all the accounts that have access to the cluster are listed e.g.
```
